domain/port/service/service_test.go

```diff
@@ -308,132 +458,403 @@ func (s *serviceSuite) TestUpdateUnitPortsClosePortConflictColocated(c *gc.C) {
+		network.MustParsePortRange("100-150/tcp"),
+	}
+
+	currentPorts := map[string][]port.PortRangeWithUUID{}
+
+	openPorts := network.GroupedPortRanges{
+		"ep1": {
+			network.MustParsePortRange("100-150/tcp"),
+		},
+	}
+
+	s.st.EXPECT().GetColocatedOpenedPorts(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(colocatedPorts, nil)
+	s.st.EXPECT().GetUnitOpenedPortsWithUUIDs(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(currentPorts, nil)
+	s.st.EXPECT().GetEndpoints(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(endpoints, nil)
+	s.st.EXPECT().AddOpenedPorts(domaintesting.IsAtomicContextChecker, map[port.UUID][]network.PortRange{
+		s.epUUID1: {
+			network.MustParsePortRange("100-150/tcp"),
+		},
+	}).Return(nil)
+
+	srv := NewService(s.st)
+	err := srv.UpdateUnitPorts(context.Background(), s.unitUUID1, openPorts, network.GroupedPortRanges{})
+	c.Assert(err, jc.ErrorIsNil)
+}
+
+func (s *serviceSuite) TestUpdateUnitPortRangesConflictsColocated(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	colocatedPorts := []network.PortRange{
+		network.MustParsePortRange("100-150/tcp"),
+	}
+
+	openPorts := network.GroupedPortRanges{
+		"ep1": {
+			network.MustParsePortRange("100/tcp"),
+		},
+	}
+
+	s.st.EXPECT().GetColocatedOpenedPorts(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(colocatedPorts, nil)
+
+	srv := NewService(s.st)
+	err := srv.UpdateUnitPorts(context.Background(), s.unitUUID1, openPorts, network.GroupedPortRanges{})
+	c.Assert(err, jc.ErrorIs, port.ErrPortRangeConflict)
+}
+
+func (s *serviceSuite) TestUpdateUnitPortRangeEqualOnOtherEndpoint(c *gc.C) {
+	defer s.setupMocks(c).Finish()
+
+	endpoints := []port.Endpoint{
+		{UUID: s.wildcardEpUUID, Endpoint: WildcardEndpoint},
+		{UUID: s.epUUID1, Endpoint: "ep1"},
+		{UUID: s.epUUID2, Endpoint: "ep2"},
+	}
+
+	currentPorts := map[string][]port.PortRangeWithUUID{
+		"ep2": {
+			{UUID: s.portRangeUUID1, PortRange: network.MustParsePortRange("100-200/tcp")},
+		},
+	}
+
+	colocatedPorts := []network.PortRange{network.MustParsePortRange("100-200/tcp")}
+
+	openPorts := network.GroupedPortRanges{
+		"ep1": {
+			network.MustParsePortRange("100-200/tcp"),
+		},
+	}
+
+	s.st.EXPECT().GetColocatedOpenedPorts(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(colocatedPorts, nil)
+	s.st.EXPECT().GetUnitOpenedPortsWithUUIDs(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(currentPorts, nil)
+	s.st.EXPECT().GetEndpoints(domaintesting.IsAtomicContextChecker, s.unitUUID1).Return(endpoints, nil)
+	s.st.EXPECT().AddOpenedPorts(domaintesting.IsAtomicContextChecker, map[port.UUID][]network.PortRange{
+		s.epUUID1: {
 			network.MustParsePortRange("100-200/tcp"),
 		},
 	}).Return(nil)
 
 	srv := NewService(s.st)
-	err := srv.UpdateUnitPorts(context.Background(), unitUUID, openPorts, closePorts)
+	err := srv.UpdateUnitPorts(context.Background(), s.unitUUID1, openPorts, network.GroupedPortRanges{})
 	c.Assert(err, jc.ErrorIsNil)
 }
```

Comment:
I want to see tests around your helper methods. I want to lock in behavior. This includes the wildcard reconcile and the filter methods.